openshift-kube-apiserver/admission/customresourcevalidation/dns/validate_dns.go

```diff
@@ -0,0 +1,203 @@
+package dns
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/validation"
+	unversionedvalidation "k8s.io/apimachinery/pkg/apis/meta/v1/validation"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kvalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/admission"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	crvalidation "k8s.io/kubernetes/openshift-kube-apiserver/admission/customresourcevalidation"
+)
+
+const PluginName = "operator.openshift.io/ValidateDNS"
+
+// Register registers the DNS validation plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return crvalidation.NewValidator(
+			map[schema.GroupResource]bool{
+				operatorv1.GroupVersion.WithResource("dnses").GroupResource(): true,
+			},
+			map[schema.GroupVersionKind]crvalidation.ObjectValidator{
+				operatorv1.GroupVersion.WithKind("DNS"): dnsV1{},
+			})
+	})
+}
+
+// toDNSV1 converts a runtime object to a versioned DNS.
+func toDNSV1(uncastObj runtime.Object) (*operatorv1.DNS, field.ErrorList) {
+	if uncastObj == nil {
+		return nil, nil
+	}
+
+	obj, ok := uncastObj.(*operatorv1.DNS)
+	if !ok {
+		return nil, field.ErrorList{
+			field.NotSupported(field.NewPath("kind"), fmt.Sprintf("%T", uncastObj), []string{"DNS"}),
+			field.NotSupported(field.NewPath("apiVersion"), fmt.Sprintf("%T", uncastObj), []string{"operator.openshift.io/v1"}),
+		}
+	}
+
+	return obj, nil
+}
+
+// dnsV1 is runtime object that is validated as a versioned DNS.
+type dnsV1 struct{}
+
+// ValidateCreate validates a DNS that is being created.
+func (dnsV1) ValidateCreate(uncastObj runtime.Object) field.ErrorList {
+	obj, errs := toDNSV1(uncastObj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	errs = append(errs, validation.ValidateObjectMeta(&obj.ObjectMeta, false, validation.NameIsDNSSubdomain, field.NewPath("metadata"))...)
```

Comment:
how many instances do you allow?  Is this not just "cluster"?

Reply:
The user can create arbitrarily many instances, but the operator only acts on the one named "default".